ts-client/src/amm/index.ts

```diff
@@ -452,6 +496,63 @@ export default class AmmImpl implements AmmImplementation {
         this.accountsInfo.currentTime,
       );
     }
+
+    // Compute pool information, Typescript implementation of https://github.com/mercurial-finance/mercurial-dynamic-amm/blob/main/programs/amm/src/lib.rs#L960
+    const currentTimestamp = new BN(this.accountsInfo.currentTime);
+    const tokenAAmount = getAmountByShare(
+      this.accountsInfo.poolVaultALp,
+      this.vaultA.getWithdrawableAmountSync(currentTimestamp.toNumber()),
+      this.accountsInfo.vaultALpSupply,
+    );
+    const tokenBAmount = getAmountByShare(
+      this.accountsInfo.poolVaultBLp,
+      this.vaultB.getWithdrawableAmountSync(currentTimestamp.toNumber()),
+      this.accountsInfo.vaultBLpSupply,
+    );
+
+    let firstTimestamp = new BN(0);
+    let apy,
+      virtualPriceNumber,
+      firstVirtualPriceNumber = 0;
+
+    const d = this.swapCurve.computeD(this.poolState.tokenAAmount, this.poolState.tokenBAmount);
+    let latestVirtualPrice = d.mul(VIRTUAL_PRICE_PRECISION).div(this.poolState.lpSupply);
+
+    if (latestVirtualPrice.eq(new BN(0))) {
+      const lastVirtualPrice = getLastVirtualPrice(this.apyState);
+      if (lastVirtualPrice) {
+        latestVirtualPrice = lastVirtualPrice.price;
+      }
+    }
+
+    const firstVirtualPrice = getFirstVirtualPrice(this.apyState);
+
+    if (firstVirtualPrice && latestVirtualPrice.gt(new BN(0))) {
+      // Compute APY
+      const second = latestVirtualPrice.toNumber() / VIRTUAL_PRICE_PRECISION.toNumber();
+      const first = firstVirtualPrice.price.toNumber() / VIRTUAL_PRICE_PRECISION.toNumber();
+      const timeElapsed = currentTimestamp.sub(firstVirtualPrice.timestamp).toNumber();
+      const rate = second / first;
+      const frequency = (365 * 24 * 3600) / timeElapsed;
+      const compoundRate = rate ** frequency;
+      apy = (compoundRate - 1) * 100;
+
+      virtualPriceNumber = second;
+      firstVirtualPriceNumber = first;
+      firstTimestamp = firstVirtualPrice.timestamp;
+    }
+
+    let poolInformation: PoolInformation = {
+      tokenAAmount,
+      tokenBAmount,
+      currentTimestamp,
+      apy,
+      firstTimestamp,
+      firstVirtualPrice: firstVirtualPriceNumber,
+      virtualPrice: virtualPriceNumber,
+    };
+
+    this.poolState = Object.assign(this.poolState, poolInformation);
```

Comment:
I think we should extract this to a private function.

Reply:
Fixed